### PR TITLE
[MCP] Skip invalidating def constant regs during forward propagation

### PR DESCRIPTION
**Pull request by vladimirradosavljevic**
Before this patch, redundant COPY couldn't be removed for the following case:
```
  %reg1 = COPY %const-reg
  ... // There is a def of %const-reg
  %reg2 = COPY killed %reg1
```
where this can be optimized to:
```
  ... // There is a def of %const-reg
  %reg2 = COPY %const-reg
```

This patch allows for such optimization by not invalidating defined constant registers. This is safe, as architectures like AArch64 and RISCV replace a dead definition of a GPR with a zero constant register for certain instructions.

### CODE REVIEW

**Comment by vladimirradosavljevic** [comment]
@qcolombet  @arsenm  @topperc  PTAL.

File: llvm/lib/CodeGen/MachineCopyPropagation.cpp [hunk 1]
```diff
@@ -885,6 +885,10 @@ void MachineCopyPropagation::ForwardCopyPropagateBlock(MachineBasicBlock &MBB) {
       assert(!Reg.isVirtual() &&
              "MachineCopyPropagation should be run after register allocation!");
 
+      // Skip invalidating constant registers.
+      if (MRI->isReserved(Reg) && MRI->isConstantPhysReg(Reg))
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineCopyPropagation.cpp:889]: Don't need the reserved check?

File: llvm/test/CodeGen/AArch64/machine-cp-constant-reg.mir [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  bb.0:
+    liveins: $w2
+    ; CHECK-LABEL: name: test
+    ; CHECK: liveins: $w2
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: dead $wzr = SUBSWri killed renamable $w2, 0, 0, implicit-def $nzcv
+    ; CHECK-NEXT: renamable $w0 = COPY $wzr
+    ; CHECK-NEXT: RET_ReallyLR implicit killed $w0
+    renamable $w1 = COPY $wzr
+    dead $wzr = SUBSWri killed renamable $w2, 0, 0, implicit-def $nzcv
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/machine-cp-constant-reg.mir:16]: Can you remove the dead flag?

File: llvm/lib/CodeGen/MachineCopyPropagation.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/CodeGen/MachineCopyPropagation.cpp:889]: This is only relevant for defs, so move this down inside the def handling?

Comment (vladimirradosavljevic) [reply llvm/lib/CodeGen/MachineCopyPropagation.cpp]: Removed reserved check and moved `isConstantPhysReg` check inside the def handling.

File: llvm/test/CodeGen/AArch64/machine-cp-constant-reg.mir (hunk 2 above)
Comment (vladimirradosavljevic) [reply llvm/test/CodeGen/AArch64/machine-cp-constant-reg.mir]: Removed dead flag.

**Review by arsenm — APPROVED** [review]

**Comment by vladimirradosavljevic** [comment]
@arsenm Thanks for the review! Could you please merge this PR on my behalf?

**Pull request merged**